### PR TITLE
[oneDNN] Improve threadpool task scheduling

### PR DESCRIPTION
**Pull request by xiangzez**
Using oneDNN with eigen threadpool has threading conflict issue that affects performance when oneDNN threadpool thread num equals available CPU core num. This change makes Tensorflow scheduling a task in caller thread to avoid this issue. The behavior is controled by an environment variable.

### CODE REVIEW

**Comment by linlifan** [comment]
@penpornk hi, penpo, could you help to review this PR? thank you

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Review by penpornk — COMMENTED** [review]
I'm so sorry for the delay! Could you please post some performance numbers with and without the changes? Thank you!

File: tensorflow/core/util/mkl_threadpool.h [hunk 1]
```diff
@@ -89,18 +91,44 @@ struct MklDnnThreadPool : public threadpool_iface {
...
+                int start, end;
+                balance211(n, njobs, i, &start, &end);
+                for (int j = start; j < end; j++) fn(j, n);
+              } else {
+                fn(i, n);
+              }
+            },
+            i, i + 1);
+      }
     }
```
Comment (penpornk) [inline tensorflow/core/util/mkl_threadpool.h:132]: We can be more concise. Let's refactor the code in the lambda function and condition the loop bound.

Refactor:
```c++
inline void RunJobs(bool balance, int i, ...) {
  if (balance) {
    int start, end;
    balance211(n, njobs, i, &start, &end);
    for (int j = start; j < end; j++) fn(j, n);
  } else {
    fn(i, n);
  }   
}
```
The code body:
```c++
    const bool use_caller_thread =
        ThreadPoolUseCallerThread() && nthr == port::NumSchedulableCPUs();
    // Schedule njobs-1 jobs to the thread pool
    const int njobs_to_schedule = use_caller_thread ? njobs - 1 : njobs;
    for (int i = 0; i < njobs_to_schedule; i++) {
      eigen_interface_->ScheduleWithHint(
          [balance, i, n, njobs, fn]() { RunJobs(balance, i, n, njobs, fn); },
          i, i + 1);
    }
    if (use_caller_thread) {
      RunJobs(balance, njobs - 1, n, njobs, fn);
    }
```

File: tensorflow/core/util/onednn_env_vars.h [hunk 2]
```diff
@@ -20,6 +20,7 @@ limitations under the License.
 namespace tensorflow {
 bool AreWeightsFrozen();
 bool UseSystemAlloc();
+bool ThreadPoolUseCallerThread();
 }  // namespace tensorflow
```
Comment (penpornk) [inline tensorflow/core/util/onednn_env_vars.h:24]: Nit: Would you mind adding an empty line before and after this line (one before and one after)?

**Review by penpornk — CHANGES_REQUESTED** [review]

**Comment by xiangzez** [comment]
Hi @penpornk, we run performance tests locally and saw average ~15% throughput improvement for FP32 models and average ~12% throughput improvement for BF16 models. The detailed test result is attached below.

[results.xlsx](https://github.com/tensorflow/tensorflow/files/10358829/results.xlsx)

File: tensorflow/core/util/mkl_threadpool.h (hunk 1 above)
Comment (xiangzez) [reply tensorflow/core/util/mkl_threadpool.h]: I have refactored the code.

**Review by penpornk — APPROVED** [review]
Thank you very much for the results and quick changes!

**Comment by penpornk** [comment]
@nSircombe @milpuz01 FYI. This PR adds an option to run job on caller thread and could potentially help with threadpool task scheduling overhead issues. It's disabled by default so it shouldn't affect aarch64 unless the env var `TF_ONEDNN_THREADPOOL_USE_CALLER_THREAD` is set.

**Comment by milpuz01** [comment]
Thanks @penpornk. Last time when we tried this patch we got segmentation fault on AArch64 build. We are now rebuilding again with master and will try it again and report on results.

If I understand correctly this patch should only help for the cases when you are using all available cores on the machines (i.e. running inference with 16 threads on 16 core machine), but it shouldn't have much impact if say running inference with 8 threads on 16 core machine as in that case there won't be any thread oversubscription?

CC: @cfRod

**Comment by penpornk** [comment]
[ARM CI failures](https://github.com/tensorflow/tensorflow/actions/runs/3856922764/jobs/6623077289) seem to be already failing tests (7 tests; no new failures).

**Comment by Srini511** [comment]
> Thanks @penpornk. Last time when we tried this patch we got segmentation fault on AArch64 build. We are now rebuilding again with master and will try it again and report on results.
> 
> If I understand correctly this patch should only help for the cases when you are using all available cores on the machines (i.e. running inference with 16 threads on 16 core machine), but it shouldn't have much impact if say running inference with 8 threads on 16 core machine as in that case there won't be any thread oversubscription?
> 
> CC: @cfRod

@milpuz01 Curious to know what the reason for the segfault is?

**Comment by nSircombe** [comment]
Hi @Srini511,

@cfRod has been looking at this in more detail and should be able to give an update. She was able to find the cause of the segfault. I think we will need to raise a follow-up PR to address that and enable for the AArch64 build.

**Comment by penpornk** [comment]
@milpuz01 @nSircombe @cfRod Sorry I forgot about the segfault. :'( 
I will review the follow-up PR as soon as it's open. 

Re: @milpuz01: 
> If I understand correctly this patch should only help for the cases when you are using all available cores on the machines (i.e. running inference with 16 threads on 16 core machine), but it shouldn't have much impact if say running inference with 8 threads on 16 core machine as in that case there won't be any thread oversubscription?

That's my understanding as well.

**Pull request merged**